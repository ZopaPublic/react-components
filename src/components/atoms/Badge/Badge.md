### Summary

Use the `<Badge />` component to label elements in the user interface.

There's four variations available and accepts the same sizes as [`<Text />`](#/Components/Atoms/Text).

### Examples

- Variations

```ts
import { Fragment } from 'react';
import { Badge } from '@zopauk/react-components';

<Fragment>
  <div>
    <Badge className="mb-6">Default</Badge>
  </div>
  <div>
    <Badge styling="waiting" className="mb-6">
      Pending
    </Badge>
  </div>
  <div>
    <Badge styling="confirmed" className="mb-6">
      Approved
    </Badge>
  </div>
  <div>
    <Badge styling="invalid">Invalid</Badge>
  </div>
</Fragment>;
```
