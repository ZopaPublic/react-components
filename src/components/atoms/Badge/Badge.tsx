import React, { HTMLAttributes } from 'react';
import styled, { css } from 'styled-components';
import { colors } from '../../../constants';
import Text from '../Text/Text';

type Styling = 'confirmed' | 'default' | 'invalid' | 'waiting' | 'brand';
type ColorMapField = { bg: string; text: string };
type ColorMap = Record<Styling, ColorMapField>;

interface BadgeProps extends Omit<HTMLAttributes<HTMLSpanElement>, 'color'> {
  /**
   * The style you want to assign to your badge.
   */
  styling?: Styling;
}

const colorMap: ColorMap = {
  brand: {
    bg: colors.brandLight,
    text: colors.greyDarkest,
  },
  confirmed: {
    bg: colors.successLight,
    text: colors.successDark,
  },
  default: {
    bg: colors.greyLighter,
    text: colors.greyDarkest,
  },
  invalid: {
    bg: colors.alertLight,
    text: colors.alertDark,
  },
  waiting: {
    bg: colors.warningLight,
    text: colors.warningDark,
  },
};

const StyledBadge = styled(Text).attrs({
  size: 'small',
  forwardedAs: 'span',
})<BadgeProps>`
  ${({ styling = 'default' }) => css`
    background-color: ${colorMap[styling].bg};
    color: ${colorMap[styling].text};
  `};

  display: inline-block;
  padding: 3px 10px;
  white-space: nowrap;
  border-radius: 12px;
`;

const Badge = ({ children, styling, ...rest }: BadgeProps) => {
  return (
    <StyledBadge styling={styling} {...rest}>
      {children}
    </StyledBadge>
  );
};

Badge.defaultProps = {
  styling: 'default',
};

export default Badge;
