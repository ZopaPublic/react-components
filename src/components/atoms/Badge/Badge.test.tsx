import axe from '../../../../axe-helper';
import React from 'react';
import { render } from '@testing-library/react';
import Badge from './Badge';

describe('<Badge />', () => {
  it('renders without crashing', () => {
    const { container } = render(<Badge>Content</Badge>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('renders with no a11y violations', async () => {
    const { container } = render(<Badge>Content</Badge>);

    expect(container.firstChild).toMatchSnapshot();
    const results = await axe(container.innerHTML);
    expect(results).toHaveNoViolations();
  });

  it('renders with specific style', () => {
    const { container } = render(<Badge styling="confirmed">Content</Badge>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
