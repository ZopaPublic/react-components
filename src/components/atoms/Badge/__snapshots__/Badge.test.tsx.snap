// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge /> renders with no a11y violations 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  text-align: inherit;
}

.c1 {
  background-color: #EFEFEF;
  color: #2C3236;
  display: inline-block;
  padding: 3px 10px;
  white-space: nowrap;
  border-radius: 12px;
}

<span
  class="c0 c1"
>
  Content
</span>
`;

exports[`<Badge /> renders with specific style 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  text-align: inherit;
}

.c1 {
  background-color: #DDFDE5;
  color: #17592B;
  display: inline-block;
  padding: 3px 10px;
  white-space: nowrap;
  border-radius: 12px;
}

<span
  class="c0 c1"
>
  Content
</span>
`;

exports[`<Badge /> renders without crashing 1`] = `
.c0 {
  margin: 0;
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  color: #2C3236;
  font-size: 14px;
  line-height: 20px;
  font-weight: 400;
  font-family: "Open Sans",Roboto,Helvetica,Arial,sans-serif;
  text-align: inherit;
}

.c1 {
  background-color: #EFEFEF;
  color: #2C3236;
  display: inline-block;
  padding: 3px 10px;
  white-space: nowrap;
  border-radius: 12px;
}

<span
  class="c0 c1"
>
  Content
</span>
`;
